drivers/cleanup.py

```diff
@@ -2532,48 +2538,58 @@ def normalizeType(type):
     return type
 
 def _gcLoop(sr, dryRun):
-    failedCandidates = []
-    while True:
-        if not sr.xapi.isPluggedHere():
-            Util.log("SR no longer attached, exiting")
-            break
-        sr.scanLocked()
-        if not sr.hasWork():
-            Util.log("No work, exiting")
-            break
-
-        if not lockRunning.acquireNoblock():
-            Util.log("Another instance already running, exiting")
-            break
-        try:
-            if not sr.gcEnabled():
+    if not lockActive.acquireNoblock():
+        Util.log("Another GC instance already active, exiting")
+        return
+    try:
+        # TODO: make the delay configurable
+        Util.log("GC active, about to go quiet")
+        time.sleep(5 * 60)
+        Util.log("GC active, quiet period ended")
+
+        while True:
+            if not sr.xapi.isPluggedHere():
+                Util.log("SR no longer attached, exiting")
                 break
-            sr.cleanupCoalesceJournals()
             sr.scanLocked()
-            sr.updateBlockInfo()
-
-            if len(sr.findGarbage()) > 0:
-                sr.garbageCollect(dryRun)
-                sr.xapi.srUpdate()
-                continue
+            if not sr.hasWork():
+                Util.log("No work, exiting")
+                break
 
-            candidate = sr.findCoalesceable()
-            if candidate:
-                util.fistpoint.activate("LVHDRT_finding_a_suitable_pair",sr.uuid)
-                sr.coalesce(candidate, dryRun)
-                sr.xapi.srUpdate()
-                continue
+            if not lockRunning.acquireNoblock():
+                Util.log("Unable to acquire GC running lock.")
+                return
+            try:
+                if not sr.gcEnabled():
+                    break
+                sr.cleanupCoalesceJournals()
+                sr.scanLocked()
```

Comment:
It's not clear that sr.cleanupCoalesceJournals doesn't require the scan to be repeated. It is a possible future improvement if this is not the case.